src/codegen/transformations.lisp

```diff
@@ -246,3 +238,173 @@
+      (make-node-abstraction
+       :type (node-type node)
+       :vars (node-abstraction-vars node)
+       :subexpr (funcall traverse
+                         (node-abstraction-subexpr node)
+                         ':bound-variables
+                         (append (node-abstraction-vars node) bound-variables))))
+    (action (:traverse node-let node traverse &key bound-variables)
+      (let ((new-args (list ':bound-variables
+                            (append
+                             (mapcar #'car (node-let-bindings node))
+                             bound-variables))))
+        (make-node-let
+         :type (node-type node)
+         :bindings (loop :for (name . node) :in (node-let-bindings node)
+                         :collect (cons name (apply traverse node new-args)))
+         :subexpr (apply traverse (node-let-subexpr node) new-args))))
+    (action (:traverse node-match node traverse &key bound-variables)
+      (make-node-match
+       :type (node-type node)
+       :expr (funcall traverse (node-match-expr node) ':bound-variables bound-variables)
+       :branches (mapcar
+                  (lambda (branch)
+                    (make-match-branch
+                     :pattern (match-branch-pattern branch)
+                     :body (funcall traverse
+                                    (match-branch-body branch)
+                                    ':bound-variables
+                                    (append (pattern-variables (match-branch-pattern branch))
+                                            bound-variables))))
+                  (node-match-branches node))))
+    (action (:traverse node-dynamic-extent node traverse &key bound-variables)
+      (let ((new-args (list ':bound-variables
+                            (cons (node-dynamic-extent-name node) bound-variables))))
+        (make-node-dynamic-extent
+         :type (node-type node)
+         :name (node-dynamic-extent-name node)
+         :node (apply traverse (node-dynamic-extent-node node) new-args)
+         :body (apply traverse (node-dynamic-extent-body node) new-args))))
+    (action (:traverse node-bind node traverse &key bound-variables)
+      (let ((new-args (list ':bound-variables
+                            (cons (node-bind-name node) bound-variables))))
+        (make-node-bind
+         :type (node-type node)
+         :name (node-bind-name node)
+         :expr (apply traverse (node-bind-expr node) new-args)
+         :body (apply traverse (node-bind-body node) new-args)))))
+   t))
+
+(defun traverse-with-binding-list (node actions)
+  "Traverse `node` while keeping track of a list of `bound-variables` which gets passed as a keyword argument whenever a function from `actions` gets called."
+  (declare (type node        node)
+           (type action-list actions)
+           (values node &optional))
+  (traverse node
+            (append actions (make-binding-list-traversals))
+            ':bound-variables nil))
```

Comment:
if this is immutable: `'(:bound-variables nil)`

Reply:
I updated `traverse` to spread `args` in every case, so any form with general `args` uses a list and `apply`. This means that a keyword argument like `bound-variables` can be put into `traverse` directly now, instead of needing to wrap it as a list.

Upon thinking more, the reason why using keyword arguments in the traversal would have been nice before is that different traversals might keep track of different data and only use some of it, but the main definition of `traverse` would have had to keep track of it all. Now that `traverse` passes arbitrary extra `args`, there can be no ambiguity, and anything using `traverse-binding-list` can be guaranteed that there is just the one arg `bound-variables`, so I changed it from a keyword argument to a regular argument.